components/hab/src/command/pkg/binlink.rs

```diff
@@ -135,89 +141,159 @@ pub fn binlink_all_in_pkg(ui: &mut UI,
     Ok(())
 }
 
-fn is_dest_on_path<T: AsRef<Path>>(dest_dir: T) -> bool {
+fn is_dest_on_path(dest_dir: &Path) -> bool {
     if let Some(val) = env::var_os("PATH") {
-        env::split_paths(&val).any(|p| p == dest_dir.as_ref())
+        env::split_paths(&val).any(|p| p == dest_dir)
     } else {
         false
     }
 }
 
+#[allow(dead_code)]
 struct Binlink {
-    dest: PathBuf,
-    src:  PathBuf,
+    dest:                PathBuf,
+    src:                 PathBuf,
+    default_interpreter: PathBuf,
 }
 
-#[cfg(not(target_os = "windows"))]
 impl Binlink {
-    pub fn new<T: AsRef<Path>>(src: T, dest_dir: T) -> Result<Self> {
-        let bin_name = match src.as_ref().file_name() {
-            Some(name) => name,
-            None => return Err(Error::CannotParseBinlinkSource(src.as_ref().to_path_buf())),
-        };
-
-        Ok(Self { dest: dest_dir.as_ref().join(bin_name),
-                  src:  src.as_ref().to_path_buf(), })
+    pub fn new(src: &Path, dest_dir: &Path, default_interpreter: &Path) -> Result<Self> {
+        Ok(Self { dest:                Self::binstub_path(&src, dest_dir)?,
+                  src:                 src.to_path_buf(),
+                  default_interpreter: default_interpreter.to_path_buf(), })
     }
 
-    pub fn from_file<T: AsRef<Path>>(dest: T) -> Result<Self> {
-        Ok(Binlink { dest: dest.as_ref().to_path_buf(),
-                     src:  fs::read_link(&dest)?, })
+    pub fn from_file(path: &Path) -> Result<Self> {
```

Comment:
unfortunately this gets mucked up with the confusion I refer to above. Because `src` throughout this code is referring to the actual `pkgs` binary and `dest` is the actual binlink like this `path` arg. Gonna think about how to make this better.

Reply:
oh ha! just now getting to your comment below where you are getting at the same point on these confusing field names.

Reply:
I think in the `Binlink` struct, I'm going to globally rename `dest` to `link_path` and `src` to `target_path`.